blog/2020-04-07-ModernJavaConstraintsNoDRL.adoc

```diff
@@ -0,0 +1,153 @@
+= Modern java constraints (no Drools Rule Language) - ConstraintStreams
+:page-interpolate: true
+:awestruct-author: ge0ffrey
+:awestruct-layout: blogPostBase
+:awestruct-tags: [feature, coding, constraint]
+:awestruct-share_image_filename: constraintStreamGroupBy.png
+
+Traditionally, to scale out with https://www.optaplanner.org/[OptaPlanner],
+you had to learn DRL. _No more._
+With the new _ConstraintsStreams_ API, inspired Java 8 Streams and SQL,
+you can now *write your constraints in Java* (or Kotlin or Scala)
+and still benefit from incremental calculation.
+
+Underneath, ConstraintsStreams (CS) still use the powerful https://www.drools.org/[Drools] engine.
+We also still fully support score DRL's too. They are _not_ deprecated.
+
+Let's start with an example.
+In nurse rostering, to avoid assigning shifts to employee `Ann`,
+you would write this constraint in DRL:
+
+[source,drl]
+----
+rule "Don't assign Ann"
+    when
+        Shift(getEmployee().getName() == "Ann")
+    then
+        scoreHolder.addSoftConstraintMatch(kcontext, -1);
+end
+----
+
+This is the same constraint in Java using ConstraintStreams:
+
+[source,java]
+----
+Constraint constraint = constraintFactory
+        .from(Shift.class)
+        .filter(shift -> shift.getEmployee().getName().equals("Ann"))
+        .penalize("Don't assign Ann", HardSoftScore.ONE_SOFT);
+----
+
+If you're familiar with SQL or Java 8 streams, this should look familiar.
+Given a potential solution with four shifts (two of which are assigned to `Ann`),
+those shifts through it flow like this:
+
+image::constraintStreamFilter_part.png[]
+
+This new approach to writing constraints has several benefits:
+
+== Incremental calculation
+
+First off, unlike an `EasyScoreCalculator`,
+ConstraintStreams still applies incremental score calculation to scale out, just like DRL.
+For example, when a move swaps the employee of two shifts, only the delta is calculated.
+That's a huge scalability gain:
+
+image::constraintStreamIncrementalCalculation.png[]
+
+== Indexing
+
+When joining multiple types, just like an SQL `JOIN` operator,
+ConstraintStreams applies hash lookups on indexes to scale better:
+
+image::constraintStreamJoinWithJoiners.png[]
+
+== IDE support
+
+Because ConstraintsStreams are written in the Java language,
+they piggy-back on very strong tooling support.
+
+Code highlighting, code completion and debugging just work:
+
+=== Code highlighting
+
+DRL code in IntelliJ IDEA Ultimate:
+
+image::codeHighlightingDRL.png[]
+
+Java code using ConstraintStreams in IntelliJ IDEA Ultimate, for the same constraints:
```

Comment:
```suggestion
Java code using Constraint Streams in IntelliJ IDEA Ultimate, for the same constraints:
```

Reply:
Personally, I wouldn't use `Ultimate` - many people won't have that. CS works the exact same way in the standard IDEA, and people should know that.

Reply:
Community doesn't have DRL support